python/cudf/cudf/core/index.py

```diff
@@ -434,6 +438,130 @@ def _concat(cls, objs):
+        """
+
+        if isinstance(other, (list, tuple)):
+            to_concat = [self] + list(other)
+        else:
+
+            this = self
+            if len(other) == 0:
+                # short-circuit and return a copy
+                to_concat = [self]
+
+            other = as_index(other)
+
+            if len(self) == 0:
+                to_concat = [other]
+
+            if len(self) and len(other):
+                if (this.dtype == "object" and other.dtype != "object") or (
+                    other.dtype == "object" and this.dtype != "object"
+                ):
+                    raise TypeError(
+                        "cudf does not support mixed types, please type-cast "
+                        "both index to same dtypes."
+                    )
+
+                if isinstance(self._values, cudf.core.column.NumericalColumn):
+                    if self.dtype != other.dtype:
+                        this, other = numeric_normalize_types(self, other)
+                to_concat = [this, other]
+
+        for obj in to_concat:
+            if not isinstance(obj, Index):
+                raise TypeError("all inputs must be Index")
+
+        return self._concat(to_concat)
+
+    def difference(self, other, sort=None):
+        """
+        Return a new Index with elements from the index that are not in
+        `other`.
+
+        This is the set difference of two Index objects.
+
+        Parameters
+        ----------
+        other : Index or array-like
+        sort : False or None, default None
+            Whether to sort the resulting index. By default, the
+            values are attempted to be sorted, but any TypeError from
+            incomparable elements is caught by cudf.
+            * None : Attempt to sort the result, but catch any TypeErrors
+              from comparing incomparable elements.
+            * False : Do not sort the result.
+
+        Returns
+        -------
+        difference : Index
+
+        Examples
+        --------
+        >>> import cudf
+        >>> idx1 = cudf.Index([2, 1, 3, 4])
+        >>> idx1
+        Int64Index([2, 1, 3, 4], dtype='int64')
+        >>> idx2 = cudf.Index([3, 4, 5, 6])
+        >>> idx2
+        Int64Index([3, 4, 5, 6], dtype='int64')
+        >>> idx1.difference(idx2)
+        Int64Index([1, 2], dtype='int64')
+        >>> idx1.difference(idx2, sort=False)
+        Int64Index([2, 1], dtype='int64')
+        """
+
+        other = as_index(other)
+
+        if self.dtype != other.dtype and (
+            cudf.utils.dtypes.is_string_dtype(self.dtype)
+            or cudf.utils.dtypes.is_string_dtype(other.dtype)
+        ):
+            if self.equals(other):
+                if cudf.utils.dtypes.is_string_dtype(self.dtype):
+                    return StringIndex([], dtype=self.dtype)
+                else:
+                    return self[:0]
+            else:
+                difference = self
```

Comment:
We only do this when either `self` xor `other` is a string dtype and we do an equals check which is expensive. Does this address a gap in logic from the anti join or is it an optimization?

Reply:
It is to bridge a gap instead of doing a try/catch around a join which will throw if we have scenario 2. 
It(line 548) is also an optimization if we consider scenario 3.

For example we have these inputs:
```python
Scenario 1
self = pd.Index(['1', '2', '3'])
other = pd.Index([1, 2, 3])

Scenario 2
self = pd.Index(['5', '6', '2', 'a', 'b', 'c'])
other = pd.Index([1.0, 5.0, 6.0, 0.0, 1.3])

Scenario 3
self = pd.Index(['1', '2', '3'])
other = pd.Index([1.0, 2.0, 3.0])

```

1. If all of the `self` were numeric/int-like values our join operation will be able to cast to int and perform a join but will result in an empty result - To avoid this 548 & 554 exists.
```python
>>> import pandas as pd
>>> s =  pd.Index(['1', '2', '3'])
>>> o = pd.Index([1, 2, 3])
>>> s.difference(o)
Index(['1', '2', '3'], dtype='object')
```

2. If some of the `self` are non-int-like values our join operation will fail and throw type-cast failure in such case also we need to return `self`:
```python
>>> s = pd.Index(['5', '6', '2', 'a', 'b', 'c'])
>>> o = pd.Index([1.0, 5.0, 6.0, 0.0, 1.3])
>>> s.difference(o)
Index(['2', '5', '6', 'a', 'b', 'c'], dtype='object')
```

3. If all of the `self` were numeric/float-like values our join operation will be able to cast to float and perform a join but will result in an empty result, this is expected  - Instead we have the equality check which returns true and we return empty Index of appropriate type. (Line 548 to 552)
```python
>>> s = pd.Index(['1', '2', '3'])
>>> o = pd.Index([1.0, 2.0, 3.0])
>>> s.difference(o)
Index([], dtype='object')
>>> o.difference(s)
Float64Index([], dtype='float64')
```

Reply:
> If all of the `self` were numeric/int-like values our join operation will be able to cast to int and perform a join but will result in an empty result - To avoid this 548 & 554 exists.

Is this the expected behavior or is this a bug? Sounds like that's unexpected behavior in join that should be fixed to me.

> If some of the `self` are non-int-like values our join operation will fail and throw type-cast failure in such case also we need to return `self`.

Sounds like in both situations 1 and 2 we can just check that if either Index's dtype is string and the other isn't then we can just short circuit and return a copy of `self` without doing any form of equality check.

> If all of the self were numeric/float-like values our join operation will be able to cast to float and perform a join but will result in an empty result, this is expected - Instead we have the equality check which returns true and we return empty Index of appropriate type. (Line 548 to 552)

How much faster are we in this situation versus doing the anti-join? I'd also like to see how much slower we are in the case where the equality check returns False and we then have to then additionally run the anti join which I would suspect is the majority of cases.

Reply:
Let me know if you need help figuring this out. I would suggest putting a breakpoint just before the actual libcudf call [here](https://github.com/rapidsai/cudf/blob/branch-0.15/python/cudf/cudf/core/join/join.py#L109) and making sure the data going into libcudf is correct at that point. If not, I would work backwards from there.

Reply:
@brandon-b-miller Thanks for the pointer, the `libcudf` code is actually performing correctly as we send in two identical float columns in scenario 1.

I think 

> 
> Sounds like in both situations 1 and 2 we can just check that if either Index's dtype is string and the other isn't then we can just short circuit and return a copy of `self` without doing any form of equality check.
> 
I think this is possible - evaluating to see if we might miss any case.


> How much faster are we in this situation versus doing the anti-join? I'd also like to see how much slower we are in the case where the equality check returns False and we then have to then additionally run the anti join which I would suspect is the majority of cases.

I'll share the benchmarks soon once I evaluate

Reply:
Okay, so after digging through our join code - which is fine.
and also digging through pandas code, looks like there is a bug in pandas: https://github.com/pandas-dev/pandas/issues/35217

Since I have implemented `Index.equals` to be in-line with pandas we are seeing this issue. 

> Sounds like in both situations 1 and 2 we can just check that if either Index's dtype is a string and the other isn't then we can just short circuit and return a copy of `self` without doing any form of equality check.

So I'll probably do this for now instead of an equality check and xfail the corresponding tests until it has been resolved on the pandas side.

But, another question is should we have the same behavior as of pandas for `floatIndex.equals` or wait until there is consistent behavior from pandas on `FloatIndex.equals`?

Reply:
I don't think it makes sense for strings and floats to ever compare to being equals and I'm okay with differing from Pandas behavior if they choose to keep that behavior. That definitely sounds like a bug to me 😄.

Reply:
Looks like there is some discussion around that too: https://github.com/pandas-dev/pandas/issues/33940